zebra-state/src/service/read/tree.rs

```diff
@@ -182,78 +150,46 @@ where
-
-    // If there's no subtrees in chain, then we have the complete list.
-    if fork_list.is_empty() {
-        return db_list;
+    let start_index = match range.start_bound().cloned() {
+        Included(start_index) | Excluded(start_index) => start_index,
+        Unbounded => panic!("range provided to orchard_subtrees() must have start bound"),
     };
 
-    // Check for inconsistent trees in the fork.
-    for (fork_index, fork_subtree) in fork_list {
-        // If there's no matching index, just update the list of trees.
-        let Some(db_subtree) = db_list.get(&fork_index) else {
-            db_list.insert(fork_index, fork_subtree);
-
-            // Stop adding new subtrees once their amount reaches the given limit.
-            if let Some(limit) = limit {
-                let subtrees_num = u16::try_from(db_list.len())
-                    .expect("There can't be more than `u16::MAX` Orchard subtrees.");
-
-                if subtrees_num == limit.0 {
+    let results = match chain.map(|chain| chain.as_ref().orchard_subtrees_in_range(range.clone())) {
+        Some(chain_results) if chain_results.contains_key(&start_index) => return chain_results,
+        Some(chain_results) => {
+            let mut db_results = db.orchard_subtree_list_by_index_for_rpc(range);
+
+            // Check for inconsistent trees in the fork.
+            for (chain_index, chain_subtree) in chain_results {
+                // If there's no matching index, just update the list of trees.
+                let Some(db_subtree) = db_results.get(&chain_index) else {
+                    db_results.insert(chain_index, chain_subtree);
+                    continue;
+                };
+
+                // We have an outdated chain fork, so skip this subtree and all remaining subtrees.
+                if &chain_subtree != db_subtree {
                     break;
                 }
+                // Otherwise, the subtree is already in the list, so we don't need to add it.
             }
 
-            continue;
-        };
-
-        // We have an outdated chain fork, so skip this subtree and all remaining subtrees.
-        if &fork_subtree != db_subtree {
-            break;
+            db_results
         }
+        None => db.orchard_subtree_list_by_index_for_rpc(range),
+    };
 
-        // Otherwise, the subtree is already in the list, so we don't need to add it.
-    }
-
-    // Make sure the amount of retrieved subtrees does not exceed the given limit.
-    #[cfg(debug_assertions)]
-    if let Some(limit) = limit {
-        assert!(db_list.len() <= limit.0.into());
-    }
-
-    // Check that we got the start subtree from the non-finalized or finalized state.
-    // (The non-finalized state doesn't do this check.)
-    if db_list.get(&start_index).is_some() {
-        db_list
+    // Check that we got the start subtree
+    if results.contains_key(&start_index) {
+        results
```

Comment:
If the start bound of `range` is `Excluded`, then we'll always return an empty map here. Should we also prohibit `Excluded` start bounds of `range`?

Reply:
If we do, let's mention it in the docs?

Reply:
It would be unusual for us to reject some range types, and only do it at runtime (and not at compile time). For every other API that accepts a range, Zebra handles all range types correctly. But I understand this is a scope increase.

See my suggestions for how to handle `Unbounded` and `Excluded` on that code, and my detailed concerns here:
https://github.com/ZcashFoundation/zebra/pull/7739#discussion_r1359946646